src/Microsoft.Health.Fhir.Core/Features/Operations/DataConvert/DataConvertRequestHandler.cs

```diff
@@ -0,0 +1,73 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using MediatR;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Core.Configs;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Features.Operations.DataConvert.Models;
+using Microsoft.Health.Fhir.Core.Features.Security;
+using Microsoft.Health.Fhir.Core.Features.Security.Authorization;
+using Microsoft.Health.Fhir.Core.Messages.DataConvert;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.DataConvert
+{
+    public class DataConvertRequestHandler : IRequestHandler<DataConvertRequest, DataConvertResponse>
+    {
+        private readonly IFhirAuthorizationService _authorizationService;
+        private readonly IDataConvertEngine _dataConvertEngine;
+        private readonly DataConvertConfiguration _dataConvertConfiguration;
+
+        public DataConvertRequestHandler(
+            IFhirAuthorizationService authorizationService,
+            IDataConvertEngine dataConvertEngine,
+            IOptions<DataConvertConfiguration> dataConvertConfiguration)
+        {
+            EnsureArg.IsNotNull(authorizationService);
+            EnsureArg.IsNotNull(dataConvertEngine);
+            EnsureArg.IsNotNull(dataConvertConfiguration);
+
+            _authorizationService = authorizationService;
+            _dataConvertEngine = dataConvertEngine;
+            _dataConvertConfiguration = dataConvertConfiguration.Value;
+        }
+
+        public async Task<DataConvertResponse> Handle(DataConvertRequest request, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(request);
+
+            if (await _authorizationService.CheckAccess(DataActions.DataConvert) != DataActions.DataConvert)
+            {
+                throw new UnauthorizedFhirActionException();
+            }
+
+            var dataConvertTask = _dataConvertEngine.Process(request);
+            return await ExecuteWithTimeout(dataConvertTask, _dataConvertConfiguration.ProcessTimeoutThreshold, cancellationToken);
+        }
+
+        private async Task<TResult> ExecuteWithTimeout<TResult>(Task<TResult> runTask, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            using (var timeoutCancellation = new CancellationTokenSource())
+            using (var combinedCancellation = CancellationTokenSource
+                .CreateLinkedTokenSource(cancellationToken, timeoutCancellation.Token))
+            {
+                var completedTask = await Task.WhenAny(runTask, Task.Delay(timeout, cancellationToken));
+                timeoutCancellation.Cancel();
+                if (completedTask == runTask)
+                {
+                    return await runTask;
+                }
+                else
+                {
+                    throw new DataConvertTimeoutException("The data convert operation has timed out.");
```

Comment:
I don't see how this cancels runTask itself.  Won't it keep running even if we throw an exception here?

Reply:
yes, the logic was not able to cancel the run task as the cancellation token is not passed to the run task. Now I have added a cancellation token in the task parameter and the task can be cancelled with either the outer canceling triggers or timeout cancelling triggers.

Reply:
Thanks.   CancellationTokenSource can take in a TimeSpan to automatically cancel when the timeout expires -- consider using that instead of Task.Delay()?